vortex-file/src/generic.rs

```diff
@@ -1,677 +1,155 @@
-                },
-                CoalescedCancellationHandle::new(
-                    Vec::new(),
-                    metrics.cancel_received.clone(),
-                    metrics.cancelled.clone(),
-                ),
-            )
-        })
-        .collect::<Vec<_>>();
+#[cfg(feature = "object_store")]
+impl VortexOpenOptions<GenericVortexFile> {
+    pub async fn open_object_store(
+        self,
+        object_store: &Arc<dyn object_store::ObjectStore>,
+        path: &str,
+    ) -> VortexResult<VortexFile> {
+        use std::path::Path;
 
-    for req in requests {
-        let idx = fetch_ranges.partition_point(|v| v.start <= req.spec.offset) - 1;
-        let (ref mut request, ref mut cancellation) = coalesced[idx];
-        request.requests.push((req.id, req.spec));
-        cancellation.push(req.cancel_handle);
-    }
+        use vortex_io::{ObjectStoreReadAt, TokioFile};
 
-    // Ensure we sort the requests by segment ID within the coalesced request.
-    for (req, _) in coalesced.iter_mut() {
-        req.requests.sort_unstable_by_key(|(id, _)| *id);
-        metrics.record(req);
-    }
-    coalesced
-}
-
-/// Returns a sorted list of ranges that cover `ranges`
-///
-/// From arrow-rs.
-fn merge_ranges<R>(ranges: R, coalesce: u64, max_read: Option<u64>) -> Vec<Range<u64>>
-where
-    R: IntoIterator<Item = Range<u64>>,
-{
-    let mut ranges: Vec<Range<u64>> = ranges.into_iter().collect();
-    ranges.sort_unstable_by_key(|range| range.start);
-
-    let mut ret = Vec::with_capacity(ranges.len());
-    let mut start_idx = 0;
-    let mut end_idx = 1;
-
-    while start_idx != ranges.len() {
-        let start = ranges[start_idx].start;
-        let mut range_end = ranges[start_idx].end;
-
-        while end_idx != ranges.len()
-            && ranges[end_idx]
-                .start
-                .checked_sub(range_end)
-                .map(|delta| delta <= coalesce)
-                .unwrap_or(true)
-        {
-            let new_range_end = range_end.max(ranges[end_idx].end);
-            if max_read.is_some_and(|max| new_range_end - start > max) {
-                break;
-            }
-            range_end = new_range_end;
-            end_idx += 1;
+        // If the file is local, we much prefer to use TokioFile since object store re-opens the
+        // file on every read. This check is a little naive... but we hope that ObjectStore will
+        // soon expose the scheme in a way that we can check more thoroughly.
+        // See: https://github.com/apache/arrow-rs-object-store/issues/259
+        let local_path = Path::new("/").join(path);
+        if local_path.exists() {
+            // Local disk is too fast to justify prefetching.
+            self.open(TokioFile::open(local_path)?).await
+        } else {
+            self.open(ObjectStoreReadAt::new(
+                object_store.clone(),
+                path.into(),
+                None,
+            ))
+            .await
```

Comment:
IMO the issue here isn't that it's "naive" but the fact that it breaks behavior Datafusion users should control.